Slack Bot Not Performing Document Search in v2.10.3 (Regression from v2.9.7)

## Summary

After upgrading from v2.9.7 to v2.10.3, the Slack bot stopped performing document searches entirely. The bot now skips knowledge-base searches and goes directly to LLM generation, resulting in "Unable to find citations to answer" errors for all substantive questions.

**Working version**: v2.9.7
**Broken version**: v2.10.3
**Impact**: Slack bot is unusable for document-based Q&A

## Environment

- **Onyx version**: v2.10.3
- **Helm chart**: 0.4.18
- **Deployment**: Kubernetes (self-hosted)
- **Database**: PostgreSQL 16 (CloudNativePG)
- **Search engine**: Vespa 8.609.39

## Expected Behaviour

When a user asks a question in Slack, the bot should:
1. Generate embeddings for the query
2. Search Vespa for relevant documents
3. Retrieve document chunks
4. Generate an answer with citations

This is how v2.9.7 behaves.

## Actual Behavior (v2.10.3)

The bot:
1. Skips document search entirely
2. Goes directly to LLM generation
3. Fails with "Unable to find citations to answer"
4. No embeddings generated, no Vespa queries performed

## Evidence

### v2.9.7 Logs (Working) ✅

```
[INFO] event=embedding_provider texts=1 chars=40 provider=EmbeddingProvider.OPENAI
[INFO] Selected ranking profile: hybrid_search_semantic_base_1536
[INFO] Retrieved 50 inference chunks for 35 documents
[INFO] event=gather_stream status=success documents_found=35
```

**Behaviour**: Bot performs a full document search pipeline before generating an answer.

### v2.10.3 Logs (Broken) ❌

```
[INFO] event=calculate_reserved_tokens prompt_tokens=1234 ...
[INFO] event=gather_stream status=success documents_found=0
[ERROR] Unable to find citations to answer
```

**Behaviour**: Bot skips search entirely, goes straight to LLM. No embedding generation, no Vespa queries, documents_found=0.

### Web Chat (v2.10.3) ✅

The web chat interface works correctly in v2.10.3 and performs searches as expected. **This is a Slack bot-specific regression.**

## Root Cause Analysis

Based on code investigation, the issue was introduced by **PR #7399** ("Refactor pre search UI backend", merged Jan 14, 2026):

### What Happened

1. **Old code (v2.9.7)**: `backend/onyx/onyxbot/slack/handlers/handle_regular_answer.py` explicitly forced search:
   ```python
   retrieval_details = RetrievalDetails(
       run_search=OptionalSearchSetting.ALWAYS,  # ← Forced search
       real_time=False,
       filters=filters,
       enable_auto_detect_filters=auto_detect_filters,
   )

   answer_request = prepare_chat_message_request(
       message_text=user_message.message,
       retrieval_details=retrieval_details,  # Passed to request
       # ... rest of config
   )
   ```

2. **Refactoring (PR #7399, merged Jan 14, 2026)**: "Refactor pre search UI backend"
   - Changed `handle_regular_answer.py` to use the new `SendMessageRequest` API
   - Removed `RetrievalDetails` with `run_search=ALWAYS`
   - Removed `prepare_chat_message_request()` function
   - Changed from `stream_chat_message_objects()` to `handle_stream_message_objects()`
   - **Failed to preserve search forcing logic for Slack bot**
   - Tests were disabled with the comment: "reenable once new search endpoint is out"

   **Diff excerpt** (lines 214-226):
   ```diff
   -        retrieval_details = RetrievalDetails(
   -            run_search=OptionalSearchSetting.ALWAYS,
   -            real_time=False,
   -            filters=filters,
   -            enable_auto_detect_filters=auto_detect_filters,
   -        )
   -
   -        with get_session_with_current_tenant() as db_session:
   -            answer_request = prepare_chat_message_request(
   +        new_message_request = SendMessageRequest(
   +            message=user_message.message,
   +            allowed_tool_ids=None,
   +            forced_tool_id=None,
   ```

3. **New code (v2.10.3)**: Uses `SendMessageRequest` without forcing search:
   ```python
   new_message_request = SendMessageRequest(
       message=user_message.message,
       allowed_tool_ids=None,
       forced_tool_id=None,  # This is NOT forcing search!
       file_descriptors=[],
       internal_search_filters=filters,
       deep_research=False,
       origin=MessageOrigin.SLACKBOT,
       chat_session_info=ChatSessionCreationRequest(
           persona_id=persona.id,
       ),
   )
   ```

### Why It Fails

In v2.10.3, the LLM decides whether to search documents (instead of always searching). For Slack bot messages, the LLM consistently chooses **NOT** to search, resulting in the regression.

The web chat works because it uses a different code path that properly handles search.

## Steps to Reproduce

1. Deploy Onyx v2.10.3 with Slack bot integration
2. Ask the bot a question that requires document search, e.g., "How do I create a patch?"
3. Observe bot responds: "Sorry, we weren't able to find anything relevant"
4. Check logs: No `event=embedding_provider`, no Vespa queries, `documents_found=0`

## Proposed Fix

Add `forced_tool_id` to force the search tool in Slack bot's `SendMessageRequest`:

**File**: `backend/onyx/onyxbot/slack/handlers/handle_regular_answer.py` (or equivalent post-refactoring file)

```python
# Get the search tool ID
with get_session_with_current_tenant() as db_session:
    all_tools = get_tools(db_session)
    search_tool_id = next(
        (tool.id for tool in all_tools if tool.in_code_tool_id == SEARCH_TOOL_ID),
        None,
    )

# Force search tool
new_message_request = SendMessageRequest(
    message=user_message.message,
    forced_tool_id=search_tool_id,  # <-- ADD THIS
    internal_search_filters=filters,
    # ... rest of config
)
```

This restores the v2.9.7 behaviour where the Slack bot always searches documents before answering.

## Related Issues

- **Issue #5760**: "Slack bot hardcodes run_search=always" - Closed as "Not Planned"
  - This issue was about the OLD code pre-refactoring
  - Ironically, the refactoring that removed the hardcoded behaviour broke search entirely
- **PR #7399**: ["Refactor pre search UI backend"](https://github.com/onyx-dot-app/onyx/pull/7399) - The refactoring that introduced this regression
  - [Full diff](https://patch-diff.githubusercontent.com/raw/onyx-dot-app/onyx/pull/7399.diff)
  - See changes to `handle_regular_answer.py` at lines 214-226 of diff

## Additional Context

### Why This Wasn't Caught

1. No tests exist for Slack bot search behaviour
2. The refactoring disabled existing tests with a comment about "new search endpoint"
3. The "new search endpoint" was never completed before the v2.10.3 release

### Impact on Users

- Slack bot is completely unusable for document Q&A
- Forces users to use the web interface instead
- Loss of key integration feature

### Workaround

Until fixed, users must either:
1. Use web chat interface (works correctly)
2. Apply a custom patch to force the search tool
3. Rollback to v2.9.7 (requires database restore)

## Versions Affected

- ✅ v2.9.7 and earlier: Working
- ❌ v2.10.3: Broken
- ❌ v2.11.1-beta: Still broken (no fix in release notes)
- ❌ v2.11.3-beta.1: Still broken (no fix in release notes)

## Request

Please restore the Slack bot's document search functionality to match v2.9.7 behaviour. The bot should always search documents before generating answers, as this is the expected and useful behaviour for a knowledge base bot.

Thank you!

P.S.: First, I upgraded Onyx from v2.8.0 to v2.10.3 and noticed the issue with the Slackbot. Tried to downgrade it to v2.8.0, but that caused conflicts between the backend and the database structure. Then I tried switching to v2.9.7, with the same effect. I didn't risk patching the source code. Thus, I've ended up with a broken Slack bot (which is the main functionality that we're using right now).

--- a/backend/onyx/onyxbot/slack/handlers/handle_regular_answer.py
+++ b/backend/onyx/onyxbot/slack/handlers/handle_regular_answer.py
@@ -34,6 +34,8 @@ from onyx.server.query_and_chat.models import (
     MessageOrigin,
     SendMessageRequest,
 )
+from onyx.tools.constants import SEARCH_TOOL_ID
+from onyx.tools.tool_implementations.search.search_tool import SearchTool
 from onyx.utils.logger import OnyxLoggingAdapter
 from onyx.utils.retry_wrapper import retry_builder
 from shared_configs.contextvars import CURRENT_USER_ID_CONTEXTVAR
@@ -331,10 +333,27 @@ def handle_regular_answer(
             tags=channel_tags if channel_tags else None,
         )
 
+        # Slack answers should be grounded in retrieval (pre-#7399 behavior):
+        # force the search tool on the first LLM cycle, but only when the
+        # channel persona actually has a usable one — non-search personas and
+        # deployments without connectors keep the unforced behavior.
+        forced_search_tool_id = next(
+            (
+                tool.id
+                for tool in persona.tools
+                if tool.in_code_tool_id == SEARCH_TOOL_ID
+            ),
+            None,
+        )
+        if forced_search_tool_id is not None and not SearchTool.is_available(
+            db_session
+        ):
+            forced_search_tool_id = None
+
         new_message_request = SendMessageRequest(
             message=user_message.message,
             allowed_tool_ids=None,
-            forced_tool_id=None,
+            forced_tool_id=forced_search_tool_id,
             file_descriptors=[],
             internal_search_filters=filters,
             deep_research=False,

--- a/backend/tests/unit/onyx/onyxbot/test_handle_regular_answer.py
+++ b/backend/tests/unit/onyx/onyxbot/test_handle_regular_answer.py
@@ -100,6 +100,7 @@ def _make_slack_channel_config(
     persona.id = 123
     persona.name = "Scoped Agent"
     persona.document_sets = []
+    persona.tools = []
 
     slack_channel_config = MagicMock()
     slack_channel_config.persona = persona
@@ -579,3 +580,78 @@ def test_private_channel_non_ephemeral_attributes_usage(
     assert mock_respond.call_args.kwargs["receiver_ids"] is None
     mock_update_react.assert_called_once()
     assert mock_update_react.call_args.kwargs["remove"] is True
+
+
+@pytest.mark.parametrize(
+    "has_search_tool,search_available,expected_forced_id",
+    [
+        (True, True, 77),  # persona has SearchTool and it is usable -> forced
+        (True, False, None),  # SearchTool attached but unavailable -> not forced
+        (False, True, None),  # persona without SearchTool -> not forced
+    ],
+)
+def test_search_tool_forced_only_when_usable(
+    has_search_tool: bool,
+    search_available: bool,
+    expected_forced_id: int | None,
+) -> None:
+    """Slack answers force the persona's search tool on the first LLM cycle
+    (regression for the #7399 refactor dropping guaranteed retrieval), but
+    never for personas without a usable search tool."""
+    slack_channel_config = _make_slack_channel_config()
+    persona = slack_channel_config.persona
+    if has_search_tool:
+        search_tool = MagicMock()
+        search_tool.id = 77
+        search_tool.in_code_tool_id = "SearchTool"
+        persona.tools = [search_tool]
+
+    user = MagicMock()
+    user.id = uuid4()
+
+    def _gather_stream(_: object) -> ChatBasicResponse:
+        return ChatBasicResponse(
+            answer="answer",
+            answer_citationless="answer",
+            top_documents=[],
+            error_msg=None,
+            message_id=1,
+            citation_info=[],
+        )
+
+    with (
+        patch(f"{_HANDLE_REGULAR_ANSWER}.get_user_by_email", return_value=user),
+        patch(f"{_HANDLE_REGULAR_ANSWER}.get_anonymous_user", return_value=MagicMock()),
+        patch(f"{_HANDLE_REGULAR_ANSWER}.get_persona_by_id", return_value=persona),
+        patch(f"{_HANDLE_REGULAR_ANSWER}.rate_limits", side_effect=_identity_decorator),
+        patch(
+            f"{_HANDLE_REGULAR_ANSWER}.retry_builder", side_effect=_identity_decorator
+        ),
+        patch(
+            f"{_HANDLE_REGULAR_ANSWER}.get_channel_name_from_id",
+            return_value=("channel", False),
+        ),
+        patch(f"{_HANDLE_REGULAR_ANSWER}.gather_stream", side_effect=_gather_stream),
+        patch(f"{_HANDLE_REGULAR_ANSWER}.build_slack_response_blocks", return_value=[]),
+        patch(
+            f"{_HANDLE_REGULAR_ANSWER}.handle_stream_message_objects",
+            return_value=iter(()),
+        ) as mock_handle_stream_message_objects,
+        patch(f"{_HANDLE_REGULAR_ANSWER}.SearchTool") as mock_search_tool_cls,
+        patch(f"{_HANDLE_REGULAR_ANSWER}.respond_in_thread_or_channel"),
+        patch(f"{_HANDLE_REGULAR_ANSWER}.update_emote_react"),
+    ):
+        mock_search_tool_cls.is_available.return_value = search_available
+        handle_regular_answer(
+            message_info=_make_slack_message_info(ChannelType.IM),
+            slack_channel_config=slack_channel_config,
+            receiver_ids=None,
+            client=MagicMock(),
+            channel="C123",
+            logger=_mock_logger(),
+            db_session=MagicMock(),
+            feedback_reminder_id=None,
+        )
+
+    stream_call_kwargs = mock_handle_stream_message_objects.call_args.kwargs
+    assert stream_call_kwargs["new_msg_req"].forced_tool_id == expected_forced_id

